### ISSUE TITLE
getElementsByClassName

### ISSUE THREAD

**Issue by alphanull** [WebReflection/ie8#22]
Would it make sense to add this?

**Comment by WebReflection** [comment]
I don't think so, you can use `document.queryAll('.class-name')` instead. I am not planning to add another live collection on  the plate, these are slow, greedy, dated, etc.

**Comment by alphanull** [comment]
Well, I think I have to disagree here:

1) You already added other DOM related functions like previousElementSibling, so I think it would fit to also add this one, its simply missing in IE8 like the other ones.
2) It would be relatively easy to implement
3) using getElementsByClassName is **way** faster than using document.queryAll in browsers that natively support it (see https://jsperf.com/getelementsbyclassname-vs-queryselectorall/25)
4) therefore if you want to use this (and still support IE8), you either have inferior performance with having to use queryAll, or end up doing some kind of browser sniffing and adding a branch - and have much more convoluted code - or you have to use *another* shim just for this case.

Wouldn't it be much more convenient to have a single shim that handles it all? So _please_ consider adding this.

**Comment by WebReflection** [comment]
Do you understand what is a live collection? I think you are not
considering what it takes to implement it in IE8, it will kill
performances, not improve them. The direction is in any case to use
.query() and .queryAll() which are polifilled for IE8 too in the DOM4 poly.

I won't waste time in here to support a live collection for a death
browser. Apologies for any inconvenience. PR with standard compliant tests
welcome.

**Comment by alphanull** [comment]
Yes I am aware what a live collection is. But I (and I guess many others as well) do simply not need a live collection behavior as I only want to get a reference to element(s) with a certain class and immediatly do something with them. So yes, there would be a difference to the native function but at least _I_ could live with that restriction.

But I understand that you do not want to implement something that is not 100% compliant, and I see that making it act as a live collection would add quite some work and also performance implications. OK, DOM4 shim also adds query() but I guess I still have to use just another shim due to performance reasons.

And yes, IE8 is surely dead, but unfortunately not as dead as I want it to be. Was shocked with a recent project to still see all those IE8s when analysing some log files :(

**Comment by WebReflection** [comment]
The thing is, the effort needed to implement, test, and maintain a possibly
broken shim out weight the benefit. Is getElementsByClassName **really**
the bottleneck, in terms of performance, for your project?

**Comment by alphanull** [comment]
Yes, of course it can hurt performance, and it imho does not even have to be *the* bottleneck. I mean, why should I use querySelector in the first place when I know that using getElementsByClassName is from about 100x up to 3000x(!) faster in a browser that natively supports it - and of course thats the majority of them. It just hurts a bit to need additional measures to cope with IE8, if a relatively simple shim would suffice, without needing it to be 100% compliant.

**Comment by WebReflection** [comment]
I wrote it badly, let me rephrase : is the absence of
getElementsByClassName **really** a bottleneck for you? I'd love to see the
rest of the code or a profiled snapshot. Microbenchmarks are not a valid
source of data. The whole picture is.

**Comment by alphanull** [comment]
I think I understand what you mean ;) And I also know the issues with micro- or premature optimization. No, its not **really** a bottleneck, but on the other hand, I have no **huuge** bottleneck at all - I think at least. Its all those little things which add up, and I tend to avoid ways to slow down my code when I know that theres a more performant way to do it. Call it micro-optimization - but in this case I think it does not hurt me in any way to go a faster route from the beginning.

**Comment by WebReflection** [comment]
Considering I won't add the method, what stops you to use a one-line non
standard `HTMLElement.prototype.getElementsByClassName` that returns
`this.querySelectorAll ('.'+className)` ?

**Comment by alphanull** [comment]
Yeah, I guess that's exactly what I will do.
